clients/pkg/logentry/stages/extensions.go

```diff
@@ -35,9 +41,50 @@ func NewDocker(logger log.Logger, registerer prometheus.Registerer) (Stage, erro
 	return NewPipeline(logger, stages, nil, registerer)
 }
 
+type cri struct {
+	// bounded buffer for CRI-O Partial logs lines (identified with tag `P` till we reach first `F`)
+	partialLines    []string
+	maxPartialLines int
+	base            *Pipeline
+}
+
+// implement Stage interface
+func (c *cri) Name() string {
+	return "cri"
+}
+
+// implements Stage interface
+func (c *cri) Run(entry chan Entry) chan Entry {
+	entry = c.base.Run(entry)
+
+	in := RunWithSkip(entry, func(e Entry) (Entry, bool) {
+		if e.Extracted["flags"] == "P" {
+			if len(c.partialLines) >= c.maxPartialLines {
+				// Merge existing partialLines
+				newPartialLine := e.Line
+				e.Line = strings.Join(c.partialLines, "\n")
+				level.Warn(c.base.logger).Log("msg", "cri stage: partial lines upperbound exceeded. merging it to single line", "threshold", MaxPartialLinesSize)
+				c.partialLines = c.partialLines[:0]
+				c.partialLines = append(c.partialLines, newPartialLine)
+				return e, false
+			}
+			c.partialLines = append(c.partialLines, e.Line)
+			return e, true
+		}
+		if len(c.partialLines) > 0 {
+			c.partialLines = append(c.partialLines, e.Line)
+			e.Line = strings.Join(c.partialLines, "\n")
```

Comment:
I know this has already been merged, but I've got a question: what's the reason for joining with a `\n` here? 

Given that we originally had a message, split by cri-o into partial lines, adding a linefeed is effectively changing the resulting message.

For example, a json log message

```text
2022-09-20T17:51:01.214529932Z stdout P {"a": "very long json field 
2022-09-20T17:51:01.214697597Z stdout F split in two lines"}
```

will result in something that cannot be handled by a json parser

```text
2022-09-20T17:51:01.214697597+0000	{stream="stdout"}	{"a": "very long json field 
split in two lines"}
```

since control characters isn't valid in a json string.

This can be "fixed" by an extra stage replacing all linefeeds, but in this case it shouldn't be needed.

Reply:
I think you are right. Let me fix that.

Reply:
Fix: https://github.com/grafana/loki/pull/7997